From eb123390c8319ce52f3903c08ad360472f0ad121 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20=C3=85dahl?= <jadahl@gmail.com>
Date: Wed, 25 Sep 2019 22:23:09 +0200
Subject: [PATCH] display: Handle late unredirect (un)inhibit calls

When tearing down, gnome-shell may call various methods a bit late,
specifically before MetaDisplay is closed, but after MetaCompositor is
freed.

Handle calls to the fullscreen unredirect inhibitation counters
happening after MetaCompositor tear down by ignoring them. We're closing
anyway, so it's not a problem.

Fixes: https://gitlab.gnome.org/GNOME/gnome-shell/issues/1710

https://gitlab.gnome.org/GNOME/mutter/merge_requests/808
---
 src/compositor/compositor.c | 21 +++++++++++++++------
 1 file changed, 15 insertions(+), 6 deletions(-)

diff --git a/src/compositor/compositor.c b/src/compositor/compositor.c
index 1ab65ed5a..b831f74d0 100644
--- a/src/compositor/compositor.c
+++ b/src/compositor/compositor.c
@@ -1285,10 +1285,14 @@ meta_compositor_class_init (MetaCompositorClass *klass)
 void
 meta_disable_unredirect_for_display (MetaDisplay *display)
 {
-  MetaCompositor *compositor = get_compositor_for_display (display);
-  MetaCompositorPrivate *priv =
-    meta_compositor_get_instance_private (compositor);
+  MetaCompositor *compositor;
+  MetaCompositorPrivate *priv;
 
+  compositor = get_compositor_for_display (display);
+  if (!compositor)
+    return;
+
+  priv = meta_compositor_get_instance_private (compositor);
   priv->disable_unredirect_count++;
 }
 
@@ -1302,9 +1306,14 @@ meta_disable_unredirect_for_display (MetaDisplay *display)
 void
 meta_enable_unredirect_for_display (MetaDisplay *display)
 {
-  MetaCompositor *compositor = get_compositor_for_display (display);
-  MetaCompositorPrivate *priv =
-    meta_compositor_get_instance_private (compositor);
+  MetaCompositor *compositor;
+  MetaCompositorPrivate *priv;
+
+  compositor = get_compositor_for_display (display);
+  if (!compositor)
+    return;
+
+  priv = meta_compositor_get_instance_private (compositor);
 
   if (priv->disable_unredirect_count == 0)
     g_warning ("Called enable_unredirect_for_display while unredirection is enabled.");
-- 
2.22.0

